src/tdastro/sources/bayeSN_model.py

```diff
@@ -0,0 +1,493 @@
+            Knot positions in y.
+        z : (n, m) array_like
+            Function values on 2D grid.
+        Mx : (n, m) array_like
+            Second derivatives w.r.t. x.
+        My : (n, m) array_like
+            Second derivatives w.r.t. y.
+        xq : (p,) array_like
+            Query positions in x.
+        yq : (q,) array_like
+            Query positions in y.
+
+        Returns
+        -------
+        Zq : (p, q) ndarray
+            Interpolated 2D surface values.
+        """
+        xq = np.atleast_1d(xq)
+        yq = np.atleast_1d(yq)
+        nx, ny = len(x), len(y)
+        px, py = len(xq), len(yq)
+
+        xi_idx = np.clip(np.searchsorted(x, xq) - 1, 0, nx - 2)
+        yj_idx = np.clip(np.searchsorted(y, yq) - 1, 0, ny - 2)
+
+        Zq = np.zeros((px, py))
+
+        for j in range(py):
+            y0_idx = yj_idx[j]
+            y1_idx = y0_idx + 1
+            hy = y[y1_idx] - y[y0_idx]
+            ay = (y[y1_idx] - yq[j]) / hy
+            by = 1 - ay
+
+            f_y0 = []
+            f_y1 = []
+
+            for i in range(px):
+                x0_idx = xi_idx[i]
+                x1_idx = x0_idx + 1
+                hx = x[x1_idx] - x[x0_idx]
+                ax = (x[x1_idx] - xq[i]) / hx
+                bx = 1 - ax
+
+                # Values at 4 corners
+                z00 = z[x0_idx, y0_idx]
+                z10 = z[x1_idx, y0_idx]
+                z01 = z[x0_idx, y1_idx]
+                z11 = z[x1_idx, y1_idx]
+
+                Mx00 = Mx[x0_idx, y0_idx]
+                Mx10 = Mx[x1_idx, y0_idx]
+                Mx01 = Mx[x0_idx, y1_idx]
+                Mx11 = Mx[x1_idx, y1_idx]
+
+                # Spline in x for fixed y0 and y1
+                fx0 = (ax**3 * Mx00 + bx**3 * Mx10) * hx / 6 + (
+                    ax * (z00 - Mx00 * hx**2 / 6) + bx * (z10 - Mx10 * hx**2 / 6)
+                )
+                fx1 = (ax**3 * Mx01 + bx**3 * Mx11) * hx / 6 + (
+                    ax * (z01 - Mx01 * hx**2 / 6) + bx * (z11 - Mx11 * hx**2 / 6)
+                )
+
+                f_y0.append(fx0)
+                f_y1.append(fx1)
+
+            f_y0 = np.array(f_y0)
+            f_y1 = np.array(f_y1)
+
+            # Spline in y using fx0 and fx1
+            My0 = My[xi_idx, y0_idx]
+            My1 = My[xi_idx, y1_idx]
+
+            Zq[:, j] = (ay**3 * My0 + by**3 * My1) * hy / 6 + (
+                ay * (f_y0 - My0 * hy**2 / 6) + by * (f_y1 - My1 * hy**2 / 6)
+            )
+
+        return Zq
+
+    def evaluate_2d_cubic_spline(self, x, y, z, xq, yq):
+        """
+        Evaluate 2D natural cubic spline at given query points with auto second derivative computation.
+
+        Parameters
+        ----------
+        x : (n,) array_like
+            Knot positions in x direction.
+        y : (m,) array_like
+            Knot positions in y direction.
+        z : (n, m) array_like
+            Function values at grid points.
+        xq : (p,) array_like
```

Comment:
I like the "(size) array_like" type annotation a lot. We should adopt this other places in the code.